### ISSUE TITLE
sqlite error

### ISSUE THREAD

**Issue by NewtonsScan** [aaronmelton/speedtest_aio#2]
I tried to fugure this out by myself, but I'm unfortunately new to Docker and Grafan and am more used to Lua and C++ then Python :(
When i try to run this as stated in another issue the Dashborad does not load completly, specifically it can't load data. 

![Bild_2022-02-18_094245](https://user-images.githubusercontent.com/53621751/154648363-84a2f508-31ae-45c5-af04-39dbffd75a8f.png)

![image](https://user-images.githubusercontent.com/53621751/154648600-2b82930e-abc8-4179-b6e6-5acccf37f215.png)

Going through the logs the speedtest_grafana container I was able to find what I think causes the problem:

`logger=sqlstore t=2022-02-18T08:02:09.9+0000 lvl=info msg="Connecting to DB" dbtype=sqlite3`

`logger=migrator t=2022-02-18T08:02:09.91+0000 lvl=info msg="Starting DB migrations"`

`logger=migrator t=2022-02-18T08:02:09.92+0000 lvl=info msg="migrations completed" performed=0 skipped=381 duration=459.934µs`

And a little later: 
`18T08:05:18.44+0000 lvl=eror msg="Could not execute query" err="unable to open database file: no such file or directory" query="SELECT\n  timestamp AS \"time\",\n  avg(download_bandwidth)\nFROM speedtest\nORDER BY timestamp"`

Here are the complete logs I was looking at. The restart request for the plugin does not go away when:
          - restarting the speedtest_grafana container
          - restarting the speedtest_aio container
          - restarting speedtest_aio-master
          - Restarting docker and the docker engine
          - restarting the entire PC

![Bild_2022-02-18_092321](https://user-images.githubusercontent.com/53621751/154645320-420e3917-4ee2-4c01-96c0-0ae6d4347b59.png)
![Bild_2022-02-18_093100](https://user-images.githubusercontent.com/53621751/154646538-ae189ce8-b315-446b-b832-e8ba3e36f43c.png)

Most other error before, such as Grafana not loading, Dashboard not accessible were able to be treated by restarting a view times (probably some kind of restart requirments for Plugins. as given on the very top of the first page)

Sadly the speedtest_aio container for me has no logged events what so ever, eventhough my guess after looking through the volume and checking the place the database is supposed to be stored the this container should have the sqlite file, but the var/lib//db directory where the database should be is empty.

![Bild_2022-02-18_094659](https://user-images.githubusercontent.com/53621751/154649046-e48daa11-ca19-44ab-94f1-1235b25c8ad6.png)

I then ended up looking at the speedtest_aio.py file and realisedt, that non of the errors defined in there and related to the database where thrown.

Since I don't know python to well the only Ideas I have at this point are: 
- The file is never run thus not creating the database
- On the webpage of sqlite it states to create a new file with squlite test.db however teting .sqlite was no problem during my testing.

I hope the data provided can help solving it (or you can tell me what I can do to fix it) :)

**Comment by aaronmelton** [comment]
@NewtonsScan Thanks for the feedback.

When the Docker container is first created, there will not be any database file, so that is likely why you didn't see it.  After the script runs for the first time, you should find it in /var/db/ (after the script has collected data to populate it).  Unless you change the cron schedule, I have it configured to run every 15 minutes.  So it will execute on the hour, 15min, 30min and 45min mark.

I think you have two options:
1. You can bring up the containers `docker-compose up -d` and wait until the next 15min mark and then check the dashboard -- at which point it should be populated.
2. You can bring up the containers `docker-compose up -d`, enter the speedtest_aio container (as you did when troubleshooting) and manually run the script one time `/usr/local/bin/python /usr/src/speedtest_aio/speedtest_aio.py`.  This should also populate the dashboard.

Please try either of these and let me know if this fixes the issue.

I'll also add a note to README.md to make this clear.  Thanks!

**Comment by NewtonsScan** [comment]
So i tried both but the first didn't work within 1.5 hrs (I also didn't change the config) and when in /usr/local/bin cd python can not be performed eventhough the directory exists. Going to try to trash the docker run and reconfigure the entire thing, maybe I made some mistake during the setup phase. But thanks for the great and fast answer :)

**Comment by aaronmelton** [comment]
@NewtonsScan   What is the Operating System you're attempting to run this on?  Also, please provide me with the output of `docker-compose --version`.

**Comment by NewtonsScan** [comment]
![Bild_2022-02-18_175528](https://user-images.githubusercontent.com/53621751/154727437-aa6ce4fd-6a1c-42cf-a388-100c03195f96.png)
 
OS is Windows 10 Version 20H2 Build 19042.1526

**Comment by aaronmelton** [comment]
@NewtonsScan Thank you for troubleshooting with me!

Hmm... The Grafana errors you're seeing is expected behavior as the database doesn't exist until speedtest_aio.py runs for the first time.

I noted there were some issues running this in a previous version of docker-compose but I don't think that's the issue here.

If you don't mind continuing, please do this:
1. Shut down the running containers and remove the volumes they created: `docker-compose down` and `docker volume prune` (Careful with that last command! It will prune all other volumes you may have.  If you don't want to do that, remove them individually).
2. Move to a temporary directory.
3. Clone the repository into that directory (making sure we're starting fresh here and not bringing along any changes you might have made along the way).
4. Edit `speedtest_aio.py` and change INFO in Line 177 (`level="INFO"`) to DEBUG and save the file.
5. Proceed with your `docker-compose build` and `docker-compose up -d`
6. Enter the running Docker container for speedtest_aio: `docker exec -it <container ID> sh`
7. Run speedtest_aio.py: `/usr/local/bin/python /usr/src/speedtest_aio/speedtest_aio.py`  Did it produce any errors?  (This should populate the database and the dashboard should work after this.)
8. Provide the output of the log file `cat /var/log/speedtest_aio_20220218.log` so we can see what it's doing.

As an aside, I've been working on a new version that uses Alpine Linux which cuts about 60MB out of the total Docker size.  I've also created an ENTRYPOINT that will run speedtest_aio.py one time before Grafana is launched so the Dashboard should contain results from the start...

**Comment by aaronmelton** [comment]
@NewtonsScan Edited my previous comment since I'm a Linux user and the commands I originally provided won't work on Windows...

**Comment by NewtonsScan** [comment]
Thanks a million. Will do that tomorrow and get to you asap, if that works for you

**Comment by NewtonsScan** [comment]
So I have done as requested and figured out, that the problem seems to be that the foulder 'python' seems to be locked (which as far as I was able to gather is a problem on the docker implementation for windows).

I found that testing for the existence of the foulder returns false (shown on the picture below) which would explain why the script is never run on my end. I will try it out on my linux virtual machine to double check that this problem is infact a docker for windows problem.

![Bild_2022-02-20_180238](https://user-images.githubusercontent.com/53621751/154854593-739f2e9c-45b5-4d75-b78d-ed6c0dfd17e3.png)

The test I did for the existance of the python foulder and

![Bild_2022-02-20_180336](https://user-images.githubusercontent.com/53621751/154854634-8a840cab-f031-444b-ba54-423cae3998a3.png)

The reason I think the foulder is locked but actually exists.

**Comment by NewtonsScan** [comment]
Okay, sorry, it took me some time to find my way around the command line again. But this is the log data:

![Bild_2022-02-20_182104](https://user-images.githubusercontent.com/53621751/154855418-b28a1fd4-fb05-4633-8749-c7915a185bfb.png)
![Bild_2022-02-20_182206](https://user-images.githubusercontent.com/53621751/154855469-bb768413-8779-4905-84ab-01bc169007f6.png)

**Comment by aaronmelton** [comment]
@NewtonsScan While inside the Docker Container run the speedtest binary and see what happens...

`/usr/bin/speedtest --accept-license -f json`

It should output the failure to the console for you.

**Comment by NewtonsScan** [comment]
That runs the script after a querry to user and retunrs the expected format

![Bild_2022-02-21_212245](https://user-images.githubusercontent.com/53621751/155021813-8a44a9d5-4aef-4770-a99a-5cecaf5442d7.png)

that permission seems to have been what was missing since running the speedtest afterwards had the intended result:

![Bild_2022-02-21_212819](https://user-images.githubusercontent.com/53621751/155022392-1069e6b7-2887-432e-ac97-3a64b770c754.png)

**Comment by NewtonsScan** [comment]
Also thanks a million for taking the time to figure things out for/with me.

I have an Idea that I don't know if it is possible to do. Do you think it is possible to create kind of a report Dashboard that analises the outcome of the tests counting how many times one the column 1,2,3 are under speciffic values to make it easier to see the relative ammounts of bad result in columns after gathering a large amount of data?

**Comment by aaronmelton** [comment]
@NewtonsScan That's unexpected behavior because the `--accept-license` flag should bypass the prompt.

I've created a new branch [alpine_linux](https://github.com/aaronmelton/speedtest_aio/tree/alpine_linux) that incorporates a number of changes:
1. No longer uses Ookla's speedtest binary so no need to accept a license.
2. Runs speedtest first-thing and populates the database before Grafana ever starts (fixing the issue you originally had)
3. Squeezes an extra 60MB+ out of the original size by switching from Debian to Alpine Linux for the speedtest_aio image.

You're welcome to try that, assuming you'd want to start over:
1. Shut down the existing containers `docker-compose down`
2. Purge your old volumes (aren't compatible with new version) `docker volume prune`
3. Clone the new branch into a new directory `git clone -b alpine_linux https://github.com/aaronmelton/speedtest_aio.git`
4. Build the new images `docker-compose build --no-cache`

As for your question -- You can use the existing "Speedtest Database" Data Source to create a new dashboard.  Or you can modify the existing dashboard to add new panels.  Your choice.

You could create a new panel with a Visualization Type of "Table" and a query like this will provide you a list of all the times and values (replace 100000000 with the threshold you want to alert below):
```
SELECT
  timestamp AS "time",
  download
FROM speedtest
WHERE download <= 100000000
ORDER BY timestamp
```

**Comment by aaronmelton** [comment]
Created Issue #4, which may or may not be related to this issue.

Although I believe this was fixed in v0.4.0 based on my experience, I'm revisiting the Grafana Dashboard code to determine where the problem lies...

**Comment by aaronmelton** [comment]
Issue fixed by #5.
